### PR TITLE
Ruby on Rails capstone project - Budget app

### PR DESCRIPTION
**Pull request by the-faizmohammad**
**Ruby on Rails capstone project - Budget app**

- Implemented tasks:
- Installed the rails app using rails new Project-Budget-App --database=postgresql.
- Installed the linters (rubocop, stylelint).
- Configured the local database server.
- Updated the README.
- Created tables users, categories and expenditures to support the app features data.
- Updated the tables with foreign_keys and indexes using add_references setting the associations between the user and a category and a user and an expenditure.
- Created a Join table expenditures_categories that can establish the many-to-many relation between the expenditures and categories that has expenditure_id and category_id.
- Updated the foerign_keys to the join table and the categories and expenditures tables.
- Run the rails db:migrate to generate and update the schema file.
- Created User, Category and Expenditure models.
- Added the associations between the models.
- Created UsersController, CategoryController and ExpenditureController controllers to use the actions.
- Created respective index views to test sample data rendering.
- Added expenditures_sum and total_budget methods in the user and category models.
- Updated routes.
- Installed devise gem for user authentication.
- updated the routes to follow the devise user login and sign-up sessions.
- Installed the cancan gem for user actions authorization.
- Updated the ability.rb file with required user access.
- Authorized the user access in the user controllers with the before_action methods.
- Updated the controllers with CRUD actions.
- Updated the views for the actions.
- Added validations for the models.
- Added styles for the app pages and features according to the wireframe provided
- Installed the rspec to the project to implement unit, integration tests.
- created folders models, features and requests to add the tests.
- Added test suites testing the app features.
- For integration tests used capybara gem.
- Updated README with live demo and video presentation.
image

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,24 @@
...
+    end
+  end
+
+  def splash_screen_controller?
+    controller_name == 'splash_screen'
+  end
+
+  # def after_sign_out_path_for(resource_or_scope)
+  #     unauthenticated_root_path
+  # end
```
Comment (OyePriscilla) [inline app/controllers/application_controller.rb]: - [ ] Please kindly remove this commented code in order to make your codebase cleaner. 👍

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+class UsersController < ApplicationController
+  def index; end
+  # def sign_out_user
+  #   sign_out(current_user)
+  #   redirect_to unauthenticated_root_path
+  # end
```
Comment (OyePriscilla) [inline app/controllers/users_controller.rb]: - [ ]  Please kindly remove this commented code in order to make your codebase cleaner. 👍

File: app/views/categories/new.html.erb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+      <div class="field">
+        <%= form.label :icon %>
+        <%= form.text_field :icon, autofocus: true, autocomplete: "icon", placeholder: "e.g. https://source.unsplash.com/user/c_v_r/100x100" %>
+      </div>
+      </div>
+
+      <div class="field save-outer">
+        <%= form.submit "Save", class: "save-button" %>
+      </div>
+  <% end %>
```
Comment (OyePriscilla) [inline app/views/categories/new.html.erb:18]: - [ ] Please kindly give the submit button a fixed position so as to enhance the user experience when we have more than three items. 👍 

## Your implementation
Untitled

File: README.md [hunk 4]
```diff
@@ -1,24 +1,191 @@
...
+## ⭐️ Show your support <a name="support"></a>
+
+If you like this project Add a star to the repository!
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
+
+
+## 🙏 Acknowledgments <a name="acknowledgements"></a>
+
+I would like to thank Microverse for providing the required lessons and tutorials for successfully completing the project.
```
Comment (OyePriscilla) [inline README.md]: - [ ] Please kindly acknowlegde [Gregoire Vella on Behance](https://www.behance.net/gregoirevella) the [author of the original design](https://www.behance.net/gallery/19759151/Snapscan-iOs-design-and-branding?tracking_source=), as required by the [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/).

File: README.md [hunk 5]
```diff
@@ -1,24 +1,191 @@
...
+
+<!-- VIDEO PRESENTATION -->
+
+## 🎥 Video Presentation
+
+ <a name="video-presentation"></a>
+
+- [Project Video presentation](https://youtu.be/azrzXWr0FvY)
+
+<p align="right">(<a href="#readme-top">back to top</a>)</p>
```
Comment (OyePriscilla) [inline README.md]: - [ ] Kindly take note that I was unable to access this video because you've made it to be private. 👍

**Review by OyePriscilla — CHANGES_REQUESTED** [review]
Hi @the-faizmohammad 👏 👏 👏 👏, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

so-close

## Highlights
Use ruby gems as software packages system.
Install Ruby on Rails framework.
Use controllers to handle requests and render empty views.
Use params from browser request in a safe way.
Use preprocessed html file with embedded Ruby code.
Use layouts and templates for shared content.
Use database migration files to maintain database schema.
Use validations for models.
Set up associations between models.
Build a webapp that requires the user to log in.
Use devise gem for authentication.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: README.md (hunk 5 above)
Comment (Abenezer-Tilahun) [reply README.md]: Fixed.

File: README.md (hunk 4 above)
Comment (Abenezer-Tilahun) [reply README.md]: Fixed.

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (Abenezer-Tilahun) [reply app/controllers/users_controller.rb]: Fixed.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (Abenezer-Tilahun) [reply app/controllers/application_controller.rb]: Fixed.

File: app/views/categories/new.html.erb (hunk 3 above)
Comment (Abenezer-Tilahun) [reply app/views/categories/new.html.erb:18]: Fixed.

File: README.md [hunk 6]
```diff
@@ -1,24 +1,191 @@
...
+
+<!-- LIVE DEMO -->
+
+## 🚀 Live Demo
+
+ <a name="live-demo"></a>
+
+- [BudgetMap App](https://rails-nxiv.onrender.com/)
+
```
Comment (Abenezer-Tilahun) [inline README.md]: - [ Optional ] Kindly make sure that your Deploy application is working as expected When I checked the link i got the following error kindly fix the issue.
image

**Review by Abenezer-Tilahun — APPROVED** [review]
Hi @the-faizmohammad  
 
Your project is complete  👍 💯   There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 👏🏻

✔️ No linter errors.
✔️ You used correct Git-Flow.
✔️ Professional README  
✔️ Good Commit Messages
✔️ Good job for Implemented Suggested changes. 
✔️ Good job for Implemented RoR capstone project - day 3.


## Optional suggestions 

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application.

**Pull request merged**